scenes/worlds/world_gen/generation_steps/populate_room/generate_candelabra.gd

```diff
@@ -0,0 +1,115 @@
+
+### -----------------------------------------------------------------------------------------------
+
+
+### Built-in Virtual Overrides --------------------------------------------------------------------
+
+### -----------------------------------------------------------------------------------------------
+
+
+### Public Methods --------------------------------------------------------------------------------
+
+### -----------------------------------------------------------------------------------------------
+
+
+### Private Methods -------------------------------------------------------------------------------
+
+func _execute_step(data : WorldData, _gen_data : Dictionary, generation_seed : int):
+	_rng.seed = generation_seed
+	
+	var all_rooms := data.get_all_rooms()
+	var valid_rooms := _get_valid_rooms(all_rooms)
+	for entry in valid_rooms:
+		if _rng.randf() >= _room_chance:
+			_handle_candelabra(data, entry)
+
+
+func _get_valid_rooms(p_array: Array) -> Array:
+	var valid_rooms := []
+	
+	for entry in p_array:
+		var room_data := entry as RoomData
+		if room_data.is_min_dimension_greater_or_equal_to(_size_threshold):
+			valid_rooms.append(room_data)
+	
+#	print("valid rooms for candelabra: %s"%[valid_rooms])
+	
+	return valid_rooms
+
+
+func _handle_candelabra(world_data: WorldData, room_data: RoomData) -> void:
+	var spawn_list := _spawn_list_resource as ObjectSpawnList
+	
+	var corners := room_data.get_corners_data()
+	for key in corners.corner_positions:
+		var corner := corners.corner_positions[key] as Vector2
+		var corner_index := world_data.get_cell_index_from_int_position(int(corner.x), int(corner.y))
+		var corner_direction := corners.get_facing_vector_for(key)
+		
+		if (
+				not world_data.is_cell_free(corner_index) 
+				or _is_corner_next_to_door(world_data, corner, corner_direction)
+		):
+			continue
+		
+		var cell_position := world_data.get_local_cell_position(corner_index)
+		var spawn_data := spawn_list.get_random_spawn_data()
+		if not spawn_data.scene_path.empty():
+			spawn_data.set_center_position_in_cell(cell_position)
+			if spawn_data.scene_path.find(UNLIT_KEYWORD) != -1:
+				spawn_data.set_random_rotation_in_all_axis(_rng)
+			else:
+				var facing_angle := corners.get_facing_angle_for(key)
+				spawn_data.set_y_rotation(facing_angle)
+			
+			world_data.set_spawn_data_to_cell(corner_index, spawn_data)
+		else:
+#			print("No candelabra to spawn in this corner: %s"%[corner_index])
+			pass
+
+
+func _is_corner_next_to_door(
+		world_data: WorldData, corner_position: Vector2, corner_direction: Vector2
+) -> bool:
+	var value := false
+	
+	for direction in [Vector2(corner_direction.x, 0), Vector2(0, corner_direction.y)]:
+		var neighbour_position = corner_position + direction
+		var neighbour_index := world_data.get_cell_index_from_int_position(
+				neighbour_position.x, neighbour_position.y
+		)
+		if world_data.get_cell_type(neighbour_index) == world_data.CellType.DOOR:
```

Comment:
A celltype of Door isn't a strong guarantee if there is a door or not in that direction, and it makes no guarantees as to whether the door is pointing in the direction of this cell, and not some completely unrelated place. For cases like this, using the wall indices to look for a EdgeType of either Wall or Empty would be preferred (if all edges around this cell are walls or empties, there are no doors present)
The Edge type is also what is used both when generating the navmesh and actually spawning in doors.